src/workerd/io/io-own.c++

```diff
@@ -33,6 +33,38 @@ void DeleteQueue::scheduleDeletion(OwnedObject* object) const {
   }
 }
 
+void DeleteQueue::scheduleAction(jsg::Lock& js, kj::Function<void(jsg::Lock&)>&& action) const {
+  if (IoContext::hasCurrent() && IoContext::current().deleteQueue.get() == this) {
+    // Same thread! No need to enqueue.
+    action(js);
+  } else {
+    auto lock = crossThreadDeleteQueue.lockExclusive();
+    KJ_IF_SOME(state, *lock) {
+      state.actions.add(kj::mv(action));
+      KJ_REQUIRE_NONNULL(state.crossThreadFulfiller)->fulfill();
+    } else {
+      // The queue was deleted, likely because the IoContet was destroyed and the
+      // DeleteQueuePtr was invalidated. We aren't going to run the action in this
+      // case and it will be dropped on the floor. If there is an active IoContext,
+      // even if it is the wrong one, we should log a warning.
+      if (IoContext::hasCurrent()) {
+        auto err = v8::Exception::Error(
+            js.str("A promise was resolved or rejected from a different request context than "
+                   "the one it was created in. However, the creating context has already been "
+                   "destroyed and continuations for that request likely will not run correctly"_kj))
+                       .As<v8::Object>();
+        jsg::check(err->Set(js.v8Context(), js.str("name"_kj), js.str("Warning"_kj)));
+        auto stack = jsg::check(err->Get(js.v8Context(), js.str("stack"_kj)));
+
+        // Safe to mutate here since we have the exclusive lock on the queue above.
```

Comment:
I'm concerned that we're calling a lot of JavaScript here while holding the DeleteQueue's lock. It seems entirely possible that we'll end up calling something else that tries to take the lock -- e.g. if we trigger GC. So I think we actually need to drop the lock before calling any V8 APIs.

This comment here implies that the lock is providing some sort of safety but I don't see what that is -- the lock only protects the DeleteQueue itself, so if we're done manipulating the DeleteQueue, we can drop it.